### PR TITLE
Added drop_columns feature; usefull for optimizing cutoffs

### PR DESCRIPTION
**Pull request by monk-04**
If an HDF5 file with larger radial cutoffs and uniform knot spacing exists, it is possible to fit a model with smaller cutoffs without regenerating the HDF5 file. This feature can be used for optimizing the radial cutoffs.

This pull request implements the above feature.

regression/optimize.py provides some useful functions for getting the appropriate columns. These columns correspond to the columns of the HDF5 file (created using larger cutoffs), which should be ignored while fitting the model to lower cutoffs.

'least_squares.WeightedLinearModel.fit_from_file' and 'least_squares.WeightedLinearModel.batched_predict' accepts new argument 'drop_columns'. Providing a list of columns to this argument results in ignoring those columns while fitting and predicting steps.

Appropriate unit tests were added to 'tests/test_optimize.py' 

- [x] Create an example for the W system demonstrating this feature

### CODE REVIEW

**Review by JasonGibsonUfl — CHANGES_REQUESTED** [review]
This is an exxcelent feature of UF3 that needs to be highlighted and easily accessible. However for it to be useful it needs to be functional for more than rmin != 0. Another point of concern is that this feature will not be apparent unless a user digs into the source code. An example notebook would benefit this feature. 

`list` in lines 365 and 490 should be `List[str]`.

The function `get_bspline_config` is too rigid and variables are defined in arbitrary locations. I worry this may lead to incorrect use. For example `trailing_trim` and `leading_trim` are defined seemingly in the middle of the functions. If for some reason a user featurizes using `trailing_trim =4` they would not know that this function does not work for the given configuration.

We need to be a bit selective in how verbose we are in method and variable names. For example: `def get_possible_lower_cutoffs(original_bspline_config):` can be reduced to `def get_cuttoffs(bspline_config)` or `def get_lower_cuttoffs(bspline_config)`

In line 158 `original_bspline_config` is stated to be `bspline_config` however this is an internal definition. `original_bspline_config` should be defined as  `bspline.BSplineBasis`

method description in line 154 should be `3-body` not `2-body`

Can you provide a more informative `ValueError` for lines 101-116. It is not clear what the failed internal check is

**Comment by monk-04** [comment]
I have a checklist in my first comment that talks about adding an example.

Regarding the comment about the code not working when rmin!=0, the code might work when rmin!=0, I just haven't tested it. We should probably keep the condition for now and later remove it when we have extensively tested it.

I will make the changes in line 365 and 490.

I can make get_bspline_config a bit more flexible like accepting different knot spacing for 2-body and 3-body term. Regarding the leading trim and trailing trim, as I said for rmin I haven't had the time to test other settings.

I will add extra checks in `get_columns_to_drop_2b` and `get_columns_to_drop_3b`, to check for trims, knot_spacing, and some other stuff.

@JasonGibsonUfl I like your suggestion for renaming the functions.

Suggestions for line 158 and 154 noted.

The `ValueError` for internal checks is redundant check. For the drop_columns feature to work the smaller cutoff needs to be a knot of the original knots with larger cutoffs.

**Comment by monk-04** [comment]
As per @JasonGibsonUfl suggestions, I have made `get_bspline_config` a bit more flexible. It now works when `rmin!=0`.

I added some extra checks to `get_columns_to_drop_2b` and `get_columns_to_drop_3b` functions.

Also added one more unit test (see `config_2` in `test_optimize.py`), which is failing with my current `get_columns_to_drop_3b` algorithm.

The algorithm only fails when the `modify_3b_cutoff` is either 4.5 or 6.3. It works for other values like for example 5.4, 7.2, 1.8, etc.

@JasonGibsonUfl @sxie22 any suggestions?

**Comment by monk-04** [comment]
The algorithm is only failing when `rmin_3b!=0`. When `rmin_3b=0`, it passes all the cases I was able to come up with.

**Comment by monk-04** [comment]
I think I have found the reason why my `drop_column` algorithm is failing. Refer to issue #89 for more details.

File: uf3/regression/optimize.py [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+    #if rmin_2b!=0:
+    #    raise ValueError("Currrent version is only tested for rmin=0")
+    
+    #if rmin_3b!=0:
+    #    raise ValueError("Currrent version is only tested for rmin=0")
+
+    if not (np.isclose((rmax_2b-rmin_2b)%knot_spacing_2b,knot_spacing_2b) or \
+            np.isclose((rmax_2b-rmin_2b)%knot_spacing_2b,0)):
+
+            raise ValueError("Provided rmax_2b does not conatin integer number of \n\
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:55]: This line is line seems to be over-indented.

File: uf3/regression/optimize.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+                knots, seperated by knot_spacing_2b")
+
+    if not (np.isclose((rmax_3b-rmin_3b)%knot_spacing_3b,knot_spacing_3b) or \
+            np.isclose((rmax_3b-rmin_3b)%knot_spacing_3b,0)):
+
+            raise ValueError("Provided rmax_3b does not conatin integer number of \n\
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:61]: This one as well.

File: uf3/regression/optimize.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+                knots, seperated by knot_spacing_3b")
+
+    if leading_trim!=0:
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:64]: missing whitespace around operator

File: uf3/regression/optimize.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        raise ValueError("Currrent version is only tested for leading_trim=0")
+
+    if trailing_trim!=3:
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:67]: missing whitespace around operator

File: uf3/regression/optimize.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+        raise ValueError("Currrent version is only tested for trailing_trim=3")
+
+    rmax_3b_double = rmax_3b*2
+    if not(np.isclose((rmax_3b_double%knot_spacing_3b),0) or \
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:71]: missing whitespace around modulo operator

File: uf3/regression/optimize.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                by knot_spacing_3b, but rmax_3b_double does not. Consider changing rmax_3b \n\
+                to "+str(rmax_3b+knot_spacing_3b))
+
+    reso_2b = round((rmax_2b - rmin_2b)/knot_spacing_2b)
+    reso_3b = round((rmax_3b - rmin_3b)/knot_spacing_3b)
+
+    reso_3b_double = round((rmax_3b_double - rmin_3b)/knot_spacing_3b)
+    
+    r_min_map = {i:rmin_2b for i in chemical_system.interactions_map[2]}
+    r_min_map.update({i:[rmin_3b,rmin_3b,rmin_3b] for i in chemical_system.interactions_map[3]})
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:83]: missing whitespace after ','

File: uf3/regression/optimize.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+
+    """
+    interaction_2b = original_bspline_config.interactions_map[2][0]
+    interaction_3b = original_bspline_config.interactions_map[3][0]
+
+    rmax_2b_poss = original_bspline_config.knots_map[interaction_2b][4:-3]
+    rmax_3b_poss = original_bspline_config.knots_map[interaction_3b][0][4:-3]#[1::2]
+
+    for i in range(rmax_2b_poss.shape[0]):
+        if rmax_2b_poss[i] not in  original_bspline_config.knots_map[interaction_2b]:
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:119]: Extra white space after `in`

File: uf3/regression/optimize.py [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    for i in range(rmax_3b_poss.shape[0]):
+        if rmax_3b_poss[i] not in original_bspline_config.knots_map[interaction_3b][1]:
+            raise ValueError("Internal check failed-->3B_1!!")
+            
+    #for i in range(rmax_3b_poss.shape[0]):
+    #    if 2*rmax_3b_poss[i] not in original_bspline_config.knots_map[interaction_3b][2]:
+    #        print(interaction_3b,rmax_3b_poss[i],2*rmax_3b_poss[i])
+    #        raise ValueError("Internal check failed-->3B_2!!")
+
+    return {"rmax_2b_poss":rmax_2b_poss, "rmax_3b_poss":rmax_3b_poss}
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:135]: missing whitespace after ':'

File: uf3/regression/optimize.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    Args:
+        original_bspline_config: bspline_config used to create the HDF5 feature file.
+            This file was produced with larger cutoff.
+        modify_2b_cutoff (float): Intended 2-body cutoff
+        knot_spacing_2b (float): knot_spacing_2b used to create the HDF5 file
+    Returns:
+        columns_to_drop_2b (list): Should be passed to drop_columns argument of
+            fit_from_file
+    """
+    if original_bspline_config.leading_trim!=0:
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:155]: missing whitespace around operator

File: uf3/regression/optimize.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
...
+
+    if original_bspline_config.trailing_trim!=3:
+        raise ValueError("Currrent version is only tested for trailing_trim=3")
+    column_names = original_bspline_config.get_column_names()
+    interaction_partitions_num = original_bspline_config.get_interaction_partitions()[0]
+    interaction_partitions_posn = original_bspline_config.get_interaction_partitions()[1]
+
+    columns_to_drop_3b = []
+    for interaction in original_bspline_config.interactions_map[3]:
+        if not modify_3b_cutoff in original_bspline_config.knots_map[interaction][0]:
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:207]: This is an unconventional use of the `not` operator. typically the syntax is `not in`

File: uf3/regression/optimize.py [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+        grid_c = np.char.multiply(grid_c,16)
+        grid_c.flat[original_bspline_config.template_mask[interaction]] = column_names[start_ind_3b:end_ind_3b]
+
+        grid_c = np.delete(grid_c,np.s_[grid_c.shape[2]-3-num_columns_to_drop_3b_double:grid_c.shape[2]-3],axis=2) #4
+        grid_c = np.delete(grid_c,np.s_[grid_c.shape[1]-3-num_columns_to_drop_3b:grid_c.shape[1]-3],axis=1) #2
+        grid_c = np.delete(grid_c,np.s_[grid_c.shape[0]-3-num_columns_to_drop_3b:grid_c.shape[0]-3],axis=0) #2
+
+        columns_to_keep = grid_c[grid_c[:] != b'0000000000000000'].astype('<U16')
+
+        columns_to_keep_index_3b = np.where(np.isin(column_names[start_ind_3b:end_ind_3b],columns_to_keep))[0]
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:235]: This variable is not used.

File: uf3/regression/optimize.py [hunk 12]
```diff
@@ -0,0 +1,241 @@
...
+    resolution_map.update({i:[reso_3b,reso_3b,reso_3b_double] for i in chemical_system.interactions_map[3]})
+
+    bspline_config = bspline.BSplineBasis(chemical_system, r_min_map=r_min_map,
+            r_max_map=r_max_map,resolution_map=resolution_map,
+            trailing_trim=trailing_trim,leading_trim=leading_trim)
+
+    return bspline_config
+
+
+def get_possible_lower_cutoffs(original_bspline_config):
```
Comment (JasonGibsonUfl) [inline uf3/regression/optimize.py:98]: I do not like this function name. I suggest removing `possible` as it makes the function's use confusing.

**Review by JasonGibsonUfl — COMMENTED** [review]
I like the changes made to this feature. My comments are mostly about styling. Most are reoccurring such as no white space after `,`.  I would suggest using `flake8` to catch these.

The other concern I have is in variable naming. For example, line 229 exceeds PEP8 line length convention by over 30 characters. This is because of the length of the variable `num_columns_to_drop_3b_double` this could be shortened to something like `n_drop_3b_long`. Additionally, abbreviations should be used consistently. In line 237 index is abbreviated to `ind` but in the proceeding line it is not abbreviated.

**Comment by monk-04** [comment]
Formated the files according to @JasonGibsonUfl suggestions and also added an example jupyter notebook

**Review by sxie22 — APPROVED** [review]

**Comment by monk-04** [comment]
Merging!

**Pull request merged**